helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -557,20 +557,24 @@ protected static class PreferenceListNodeComparator implements Comparator<String
     protected final Map<String, String> _currentStateMap;
     protected final StateModelDefinition _stateModelDef;
     protected final List<String> _preferenceList;
+    ResourceControllerDataProvider _cache;
 
-    public PreferenceListNodeComparator(Map<String, String> currentStateMap,
-        StateModelDefinition stateModelDef, List<String> preferenceList) {
+    public PreferenceListNodeComparator(Map<String, String> currentStateMap, StateModelDefinition stateModelDef,
+        List<String> preferenceList, ResourceControllerDataProvider cache) {
```

Comment:
We cannot directly change the public constructor in open source.

You can add another constructor and keep the original one. So just pass null for the cache and invoke the new constructor for the old one like:
this (xxxxxx).


This can help you do the compute control as well. Means when cache is null, skip the MZ sorting and keep the old logic.